File: src/page2/Problem76.cpp
#include <stdio.h>
#define MIN(A, B) ((A) < (B) ? (A) : (B))

int main(int argc, char *argv[]) {
	if (argc != 3) {
		printf("Invalid arguments\n");
		return 1;
	}

	int number = atoi(argv[1]);
	int highest = atoi(argv[2]);

	if (number <= 0 || highest <= 0) {
		printf("Invalid arguments\n");
		return 2;
	}
	printf("%i\n", numSums(number, highest));
}

//number > 0 and highest > 0
int numSums(int number, int highest) {
	if (number == 1 || highest == 1) {
		return 1;
	}

	int sum = 1;
	int first, multiple, remainder;
	for (first = 2; first <= highest; first++) {
		if (number % first == 0) {
			sum++;
		}
		for (multiple = 1; multiple <= (number-1)/first; multiple++) {
			remainder = number - multiple*first;
			sum += numSums(remainder, MIN(first-1, remainder));
		}
	}

	return sum;
}

File: src/page2/Problem76.java
package page2;

public class Problem76 {

	public static int[][] partition = new int[101][101];
	
	public static void main(String args[])
	{
		System.out.println(numSums(100, 99));
	}
	
	static int numSums(int number, int highestPart)
	{
		if (number < 0 || highestPart <= 0) {
			return 0;
		}
		if (number == 0) {
			partition[number][highestPart] = 1;
			return 1;
		}
		if (partition[number][highestPart] != 0) {
			return partition[number][highestPart];
		}
		
		int sum = numSums(number - highestPart, highestPart) + numSums(number, highestPart - 1);
		
		partition[number][highestPart] = sum;
		return sum;
	}
}
